kernel-rs/src/pool.rs

```diff
@@ -96,107 +152,99 @@ impl<T: 'static, const CAPACITY: usize> Pool for Spinlock<RcPool<T, CAPACITY>> {
-            let entry = &mut *rc.ptr;
-
-            entry.ref_cnt -= 1;
-            if entry.ref_cnt == 0 {
-                Some(entry.data.read())
-            } else {
-                None
-            }
+    unsafe fn dealloc(&mut self, rc: UntaggedRc<T>) -> Option<T> {
+        let entry = &mut *rc.ptr;
+        entry.ref_cnt -= 1;
+
+        let val = if entry.ref_cnt == 0 {
+            Some(entry.data.read())
+        } else {
+            None
         };
 
         mem::forget(rc);
-
-        // Drop AFTER the pool lock is released, as dropping val may cause the current thread sleep.
-        mem::drop(val);
+        val
     }
 }
 
-/// A zero-sized reference of the `Pool`, represented in a type.
-///
-/// Ensures the safety of `dealloc` by PoolRef type parameter of TaggedBox. See
-/// https://ferrous-systems.com/blog/zero-sized-references/ for details.
-///
-/// # Safety
-///
-/// There should be at most one implementation of PoolRef for each Pool.
-pub unsafe trait PoolRef: Sized {
-    type Target: Pool + 'static;
+pub trait Tag: Clone + 'static {
+    type Target: Arena;
+    type Result: DerefMut<Target = Self::Target>;
 
-    fn deref() -> &'static Self::Target;
+    fn arena(&self) -> Self::Result;
+}
 
-    fn alloc(
-        val: <Self::Target as Pool>::Data,
-    ) -> Option<TaggedBox<Self, <Self::Target as Pool>::Data>> {
-        let alloc = Self::deref().alloc(val)?;
-        Some(TaggedBox {
-            alloc: ManuallyDrop::new(alloc),
-            _marker: PhantomData,
-        })
-    }
+pub struct Rc<T: Tag> {
+    tag: T,
+    inner: ManuallyDrop<<<T as Tag>::Target as Arena>::Handle>,
+}
+
+impl<A: 'static, T: Tag<Target = RcArena<A, C>>, const C: usize> Deref for Rc<T> {
+    type Target = A;
 
-    fn dealloc(tbox: TaggedBox<Self, <Self::Target as Pool>::Data>) {
-        mem::drop(tbox);
+    fn deref(&self) -> &Self::Target {
+        self.inner.deref()
     }
 }
 
-/// Allocation from `P`.
-#[repr(transparent)]
-pub struct TaggedBox<P: PoolRef, T: 'static>
-where
-    P::Target: Pool<Data = T>,
-{
-    alloc: ManuallyDrop<<P::Target as Pool>::PoolBox>,
-    _marker: PhantomData<P>,
+impl<A: 'static, T: Tag<Target = RcArena<A, C>>, const C: usize> DerefMut for Rc<T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        self.inner.deref_mut()
+    }
 }
 
-impl<P: PoolRef, T: 'static> TaggedBox<P, T>
-where
-    P::Target: Pool<Data = T>,
-{
-    /// # Safety
-    ///
-    /// `pbox` must be allocated from `P`.
-    pub unsafe fn from_unchecked(pbox: <P::Target as Pool>::PoolBox) -> Self {
-        Self {
-            alloc: ManuallyDrop::new(pbox),
-            _marker: PhantomData,
+impl<T: Tag> Drop for Rc<T> {
+    fn drop(&mut self) {
+        // SAFETY: We can ensure the box is allocated from `self.tag` by the invariant of `Tag`.
+        unsafe {
+            self.tag
+                .arena()
+                .dealloc(ManuallyDrop::take(&mut self.inner));
         }
```

Comment:
```suggestion
        let val = unsafe {
            self.tag
                .arena()
                .dealloc(ManuallyDrop::take(&mut self.inner));
        };
        
        // Drop AFTER the arena guard is dropped, as dropping val may cause the current thread sleep.
        mem::drop(val);
```
이렇게 수정하시면 패닉이 일어나지 않습니다. 제가 이전에 밟은 버그인데요, spinlockguard 를 연 상태로 file 을 drop 하면 그 내부에서 sleep을 해서 발생하는 오류입니다.